docs/man/man3/AEFileReader.3
.TH "AEFileReader" 3 "Mon Mar 18 2024 18:44:24" "Version v0.0.8.5a" "ArtyK's Console Engine" \" -*- nroff -*-
.ad l
.nh
.SH NAME
AEFileReader \- ArtyK's Engine File Reader; umm, it reads data from the given file\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fR#include <AEFileReader\&.hpp>\fP
.PP
Inherits \fB__AEModuleBase< AEFileReader >\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBAEFileReader\fP (\fBconst\fP std::string_view \fBfname\fP)"
.br
.RI "Class constructor -- constructs the instance and opens the file of given name\&. "
.ti -1c
.RI "\fBAEFileReader\fP (\fBvoid\fP) \fBnoexcept\fP"
.br
.RI "Class constructor -- constructs the instance with default values, and doesn't open the file\&. "
.ti -1c
.RI "\fBAEFileReader\fP (\fBAEFileReader\fP &&\fBfr\fP) \fBnoexcept\fP"
.br
.RI "Class move constructor\&. "
.ti -1c
.RI "\fBAEFileReader\fP (\fBconst\fP \fBAEFileReader\fP &)=\fBdelete\fP"
.br
.RI "Deleted copy constructor\&. "
.ti -1c
.RI "\fBAEFileReader\fP & \fBoperator=\fP (\fBconst\fP \fBAEFileReader\fP &)=\fBdelete\fP"
.br
.RI "Deleted copy assignment operator\&. "
.ti -1c
.RI "\fB~AEFileReader\fP (\fBvoid\fP)"
.br
.RI "Default Destructor\&. "
.ti -1c
.RI "\fBcint\fP \fBopenFile\fP (\fBconst\fP std::string_view \fBfname\fP)"
.br
.RI "Opens the file for reading\&. "
.ti -1c
.RI "\fBcint\fP \fBcloseFile\fP (\fBvoid\fP) \fBnoexcept\fP"
.br
.RI "Closes the currently opened file\&. "
.ti -1c
.RI "template<typename T > \fBcint\fP \fBreadVariable\fP (T &var) noexcept"
.br
.RI "Reads the bytes (size of T) from the file and dumps/reinterprets them as the value in the passed variable\&. "
.ti -1c
.RI "\fBcint\fP \fBreadString\fP (std::string &\fBstr\fP, \fBconst\fP std::size_t \fBdcount\fP)"
.br
.RI "Reads the given amount of bytes and interprets them as the string\&. "
.ti -1c
.RI "\fBcint\fP \fBreadString\fP (std::vector< char > &\fBstr\fP, \fBconst\fP std::size_t \fBdcount\fP)"
.br
.RI "Reads the given amount of bytes and interprets them as the string\&. "
.ti -1c
.RI "\fBcint\fP \fBreadString\fP (char *\fBstr\fP, \fBconst\fP std::size_t \fBdcount\fP) \fBnoexcept\fP"
.br
.RI "Reads the given amount of bytes and interprets them as the string\&. "
.ti -1c
.RI "\fBcint\fP \fBreadStringNL\fP (std::string &\fBstr\fP, \fBconst\fP \fBint\fP \fBdcount\fP)"
.br
.RI "Reads the characters to a std::string untill a newline (included), or untill the maximum character amount -- whichever comes first\&. "
.ti -1c
.RI "\fBcint\fP \fBreadStringNL\fP (std::vector< char > &\fBstr\fP, \fBconst\fP \fBint\fP \fBdcount\fP)"
.br
.RI "Reads the characters to a std::vector<char> untill a newline (included), or untill the maximum character amount -- whichever comes first\&. "
.ti -1c
.RI "\fBcint\fP \fBreadStringNL\fP (char *\fBstr\fP, \fBconst\fP \fBint\fP \fBdcount\fP) \fBnoexcept\fP"
.br
.RI "Reads the characters to a c-string untill a newline (included), or untill the maximum character amount -- whichever comes first\&. "
.ti -1c
.RI "\fBcint\fP \fBreadStringNULL\fP (std::string &\fBstr\fP, \fBconst\fP \fBllint\fP \fBdcount\fP)"
.br
.RI "Reads the characters to a std::string untill a NULL character(I guess, you can say 'included'?), or untill the maximum character amount -- whichever comes first\&. "
.ti -1c
.RI "\fBcint\fP \fBreadStringNULL\fP (std::vector< char > &\fBstr\fP, \fBconst\fP \fBllint\fP \fBdcount\fP)"
.br
.RI "Reads the characters to a std::string untill a NULL character(I guess, you can say 'included'?), or untill the maximum character amount -- whichever comes first\&. "
.ti -1c
.RI "\fBcint\fP \fBreadStringNULL\fP (char *\fBstr\fP, \fBconst\fP \fBllint\fP \fBdcount\fP) \fBnoexcept\fP"
.br
.RI "Reads the characters to a c-string untill a NULL character(I guess, you can say 'included'?), or untill the maximum character amount -- whichever comes first\&. "
.ti -1c
.RI "template<\fBtypename\fP T > 
.br
requires (std::is_integral<T>\fB::value\fP == \fBtrue\fP)\fBcint\fP \fBreadInt\fP (T &\fBnum\fP) \fBnoexcept\fP"
.br
.RI "Reads the bytes (size of T) from the file and dumps/reinterprets them as the value in the passed integer\&. "
.ti -1c
.RI "template<\fBtypename\fP T > 
.br
requires (std::is_integral<T>\fB::value\fP == \fBtrue\fP)\fBcint\fP \fBreadIntString\fP (T &\fBnum\fP) \fBnoexcept\fP"
.br
.RI "Reads the string of numbers from the file and convert it to int\&. "
.ti -1c
.RI "template<\fBtypename\fP T > \fBcint\fP \fBreadFloat\fP (T &\fBnum\fP) \fBnoexcept\fP(std::is_floating_point< T >\fB::value\fP==\fBtrue\fP)"
.br
.RI "Reads the bytes (size of T) from the file and dumps/reinterprets them as the value in the passed float\&. "
.ti -1c
.RI "template<\fBtypename\fP T > 
.br
requires (std::is_floating_point<T>\fB::value\fP == \fBtrue\fP)\fBcint\fP \fBreadFloatString\fP (T &\fBnum\fP) \fBnoexcept\fP"
.br
.RI "Reads the string of numbers from the file and convert it to float\&. "
.ti -1c
.RI "\fBcint\fP \fBreadBool\fP (bool &\fBnum\fP) \fBnoexcept\fP"
.br
.RI "Reads the byte from the file and dumps/reinterprets it as the value in the passed bool (null is false, anything else is true)\&. "
.ti -1c
.RI "\fBcint\fP \fBreadBoolString\fP (bool &\fBnum\fP) \fBnoexcept\fP"
.br
.RI "Reads the string in the file, looking for 'true' or 'false' (case insensitive)\&. "
.ti -1c
.RI "\fBcint\fP \fBreadChar\fP (\fBsigned\fP char &\fBnum\fP) \fBnoexcept\fP"
.br
.RI "Reads the character from the file and writes you to a given char\&. "
.ti -1c
.RI "\fBcint\fP \fBreadByte\fP (\fBunsigned\fP char &\fBnum\fP) \fBnoexcept\fP"
.br
.RI "Reads the byte from the file and writes you to a given unsigned\&. "
.ti -1c
.RI "\fBcint\fP \fBreadBytes\fP (std::vector< \fBunsigned\fP char > &\fBcdata\fP, \fBconst\fP std::size_t \fBdcount\fP)"
.br
.RI "Reads the bytes from the file and dumps them into the std::vector<unsigned char>\&. "
.ti -1c
.RI "\fBcint\fP \fBreadBytes\fP (\fBvoid\fP *\fBcdata\fP, \fBconst\fP std::size_t \fBdcount\fP) \fBnoexcept\fP"
.br
.RI "Reads the bytes from the file and dumps them into the given pointer\&. "
.ti -1c
.RI "\fBcint\fP \fBreadData_ptr\fP (\fBvoid\fP *\fBcdata\fP, \fBconst\fP std::size_t \fBdcount\fP, \fBconst\fP std::size_t \fBdsize\fP=\fBsizeof\fP(char)) \fBnoexcept\fP"
.br
.RI "Reads an array of dcount elements, each one with a size of dsize bytes, to memory block pointed by cdata\&. "
.ti -1c
.RI "bool \fBisOpen\fP (\fBvoid\fP) \fBconst\fP \fBnoexcept\fP"
.br
.RI "Checks if a file is open by this file-reader\&. "
.ti -1c
.RI "bool \fBisClosed\fP (\fBvoid\fP) \fBconst\fP \fBnoexcept\fP"
.br
.RI "Checks if a file isn't open by this file-reader\&. "
.ti -1c
.RI "\fBllint\fP \fBgetFileSize\fP (\fBvoid\fP) \fBconst\fP \fBnoexcept\fP"
.br
.RI "Returns size of the file in bytes\&. "
.ti -1c
.RI "std::string \fBgetFullFileName\fP (\fBvoid\fP) \fBconst\fP \fBnoexcept\fP"
.br
.RI "Returns the full name of currently open file, including the relative path, if such was provided\&. "
.ti -1c
.RI "std::string \fBgetRelativePath\fP (\fBvoid\fP) \fBconst\fP"
.br
.RI "Returns the relative path of the opened file\&. "
.ti -1c
.RI "std::string \fBgetFileName\fP (\fBvoid\fP) \fBconst\fP"
.br
.RI "Returns the name of the opened file, devoid of the path\&. "
.ti -1c
.RI "std::string \fBgetFullPath\fP (\fBvoid\fP) \fBconst\fP"
.br
.RI "Returns the full, absolute path of the opened file\&. "
.ti -1c
.RI "std::FILE * \fBgetFilePtr\fP (\fBvoid\fP) \fBconst\fP \fBnoexcept\fP"
.br
.RI "Returns the file pointer of this file-reader\&. "
.ti -1c
.RI "\fBllint\fP \fBgetCursorPos\fP (\fBvoid\fP) \fBconst\fP \fBnoexcept\fP"
.br
.RI "Returns current read cursor position\&. "
.ti -1c
.RI "\fBint\fP \fBsetCursorPos\fP (\fBconst\fP \fBllint\fP \fBpos\fP, \fBconst\fP \fBint\fP \fBorigin\fP=\fBSEEK_CUR\fP) \fBconst\fP \fBnoexcept\fP"
.br
.RI "Sets read cursor position to pos from origin\&. "
.ti -1c
.RI "\fBullint\fP \fBgetTotalReads\fP (\fBvoid\fP) \fBconst\fP \fBnoexcept\fP"
.br
.RI "Returns total reader requests made to file\&. "
.ti -1c
.RI "std::size_t \fBgetLastReadAmount\fP (\fBvoid\fP) \fBconst\fP \fBnoexcept\fP"
.br
.RI "Gets the last read amount of bytes from the opened file\&. "
.ti -1c
.RI "bool \fBisEOF\fP (\fBvoid\fP) \fBconst\fP \fBnoexcept\fP"
.br
.RI "Checks if the file (cursor) has reached it's (file's) end (feof) "
.ti -1c
.RI "bool \fBisError\fP (\fBvoid\fP) \fBconst\fP \fBnoexcept\fP"
.br
.RI "Checks if error was encountered while reading the file data (ferror) "
.ti -1c
.RI "\fBconstexpr\fP \fBconst\fP \fBullint\fP \fBgetModuleID\fP (\fBvoid\fP) \fBnoexcept\fP"
.br
.RI "Get the \fBid number of the instance\fP of the module\&. "
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "\fBstatic\fP \fBullint\fP \fBgetModuleAmount\fP (\fBvoid\fP) \fBnoexcept\fP"
.br
.RI "Get the \fBamount of the currently-alive instances of module T\fP\&. "
.ti -1c
.RI "\fBstatic\fP \fBconstexpr\fP \fBconst\fP std::string_view \fBgetModuleName\fP (\fBvoid\fP) \fBnoexcept\fP"
.br
.RI "Get the \fBname of the module\fP\&. "
.ti -1c
.RI "\fBstatic\fP \fBconstexpr\fP \fBconst\fP \fBullint\fP \fBgetInstanceCounter\fP (\fBvoid\fP) \fBnoexcept\fP"
.br
.RI "Get the current value of the \fBinstance counter\fP of the module\&. "
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "\fBconst\fP \fBullint\fP \fBm_ullModuleId\fP"
.br
.RI "The ID number of the current module\&. "
.in -1c
.SS "Static Protected Attributes"

.in +1c
.ti -1c
.RI "\fBstatic\fP std::atomic< std::size_t > \fBm_ullModulesAlive\fP"
.br
.RI "Amount of the currently-alive (not destroyed) instances of the module\&. "
.ti -1c
.RI "\fBstatic\fP std::atomic< \fBullint\fP > \fBm_ullInstanceCounter\fP"
.br
.RI "Amount of the instances of the module ever created during the program's runtime\&. "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "std::string \fBm_sFilename\fP"
.br
.RI "Full filename and relative path\&. "
.ti -1c
.RI "std::atomic< \fBullint\fP > \fBm_ullTotalReads\fP"
.br
.RI "Counter for total read operations for file\&. "
.ti -1c
.RI "std::size_t \fBm_szLastReadAmount\fP"
.br
.RI "The amount of read bytes during last operation\&. "
.ti -1c
.RI "std::FILE * \fBm_fpFilestr\fP"
.br
.RI "Object for file reading\&. "
.in -1c
.SH "Detailed Description"
.PP 
ArtyK's Engine File Reader; umm, it reads data from the given file\&. 

It is a wrapper around the C's FILE api, for speed and convenience\&. It can read strings (untill: max string size, newline, NULL character), bools, ints, and floats
.PP
Just create it, read stuff, and dump the terabytes of data from the file to your memory and what not\&. Hungarian notation is fr\&. (m_frMyFileReader) Flags start with AEFR_
.PP
\fBWarning\fP
.RS 4
This is not thread-safe!
.RE
.PP
\fBTodo\fP
.RS 4
add generic read function 
.RE
.PP

.PP
Definition at line \fB87\fP of file \fBAEFileReader\&.hpp\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "AEFileReader::AEFileReader (\fBconst\fP std::string_view fname)\fR [explicit]\fP"

.PP
Class constructor -- constructs the instance and opens the file of given name\&. 
.PP
\fBSee also\fP
.RS 4
\fBAEFileReader::openFile()\fP
.RE
.PP
\fBParameters\fP
.RS 4
\fIfname\fP The file name to opens
.RE
.PP

.PP
Definition at line \fB14\fP of file \fBAEFileReader\&.cpp\fP\&.
.PP
References \fBopenFile()\fP\&.
.SS "AEFileReader::AEFileReader (\fBvoid\fP)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Class constructor -- constructs the instance with default values, and doesn't open the file\&. 
.PP
Definition at line \fB102\fP of file \fBAEFileReader\&.hpp\fP\&.
.SS "AEFileReader::AEFileReader (\fBAEFileReader\fP && fr)\fR [noexcept]\fP"

.PP
Class move constructor\&. 
.PP
Definition at line \fB20\fP of file \fBAEFileReader\&.cpp\fP\&.
.SS "AEFileReader::AEFileReader (\fBconst\fP \fBAEFileReader\fP &)\fR [delete]\fP"

.PP
Deleted copy constructor\&. There is no need to copy AEFR, since access to file is in instance's FILE pointer\&. If in original instance, the file gets closed, the pointer is invalidated\&. Which can lead to\&.\&.\&.bad consequences using it again in the copied instance\&. 
.SS "AEFileReader::~AEFileReader (\fBvoid\fP)\fR [inline]\fP"

.PP
Default Destructor\&. Just flushes and closes the file\&. 
.PP
Definition at line \fB129\fP of file \fBAEFileReader\&.hpp\fP\&.
.PP
References \fBcloseFile()\fP\&.
.SH "Member Function Documentation"
.PP 
.SS "\fBAEFileReader\fP & \fBAEFileReader::operator\fP= (\fBconst\fP \fBAEFileReader\fP &)\fR [delete]\fP"

.PP
Deleted copy assignment operator\&. There is no need to copy AEFR, since access to file is in instance's FILE pointer\&. If in original instance, the file gets closed, the pointer is invalidated\&. Which can lead to\&.\&.\&.bad consequences using it again in the copied instance\&. 
.SS "\fBcint\fP AEFileReader::openFile (\fBconst\fP std::string_view fname)"

.PP
Opens the file for reading\&. 
.PP
\fBParameters\fP
.RS 4
\fIfname\fP The name of the file to open
.RE
.PP
\fBReturns\fP
.RS 4
AEFR_ERR_NOERROR if file was opened successfully; otherwise AEFR_ERR_OPEN_* flags (like AEFR_ERR_OPEN_FILE_NAME_EMPTY)
.RE
.PP

.PP
Definition at line \fB32\fP of file \fBAEFileReader\&.cpp\fP\&.
.PP
References \fBAEFR_ERR_NOERROR\fP, \fBAEFR_ERR_OPEN_FILE_ALREADY_OPENED\fP, \fBAEFR_ERR_OPEN_FILE_DOESNT_EXIST\fP, \fBAEFR_ERR_OPEN_FILE_NAME_EMPTY\fP, \fBace::utils::fopenCC()\fP, \fBisClosed()\fP, \fBisOpen()\fP, \fBm_fpFilestr\fP, \fBm_sFilename\fP, \fBm_szLastReadAmount\fP, and \fBreadVariable()\fP\&.
.SS "\fBcint\fP AEFileReader::closeFile (\fBvoid\fP)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Closes the currently opened file\&. 
.PP
\fBReturns\fP
.RS 4
AEFR_ERR_NOERROR if file was closed successfully; AEFR_ERR_FILE_NOT_OPEN if file isn't open
.RE
.PP

.PP
Definition at line \fB146\fP of file \fBAEFileReader\&.hpp\fP\&.
.PP
References \fB_AEFR_EXIT_ON_CLOSED_FILE\fP, \fBAEFR_ERR_NOERROR\fP, \fBm_fpFilestr\fP, \fBm_sFilename\fP, and \fBm_szLastReadAmount\fP\&.
.SS "template<typename T > \fBcint\fP AEFileReader::readVariable (T & var)\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Reads the bytes (size of T) from the file and dumps/reinterprets them as the value in the passed variable\&. 
.PP
\fBNote\fP
.RS 4
If EOF/Error was encountered when reading, the rest of bytes that weren't filled are zeroed
.RE
.PP
\fBTemplate Parameters\fP
.RS 4
\fIT\fP The type of the variable passed
.RE
.PP
\fBParameters\fP
.RS 4
\fIvar\fP The variable to fill with bytes
.RE
.PP
\fBReturns\fP
.RS 4
AEFR_ERR_READ_SUCCESS on success; otherwise AEFR_ERR_FILE_NOT_OPEN if file isn't open, or other AEFR_ERR_READ_* flags (like AEFR_ERR_READ_EOF)
.RE
.PP

.PP
Definition at line \fB166\fP of file \fBAEFileReader\&.hpp\fP\&.
.PP
References \fB_AEFR_EXIT_ON_READ_CLOSED_FILE\fP, \fBAENULL\fP, and \fBreadData_ptr()\fP\&.
.SS "\fBcint\fP AEFileReader::readString (std::string & str, \fBconst\fP std::size_t dcount)\fR [inline]\fP"

.PP
Reads the given amount of bytes and interprets them as the string\&. 
.PP
\fBNote\fP
.RS 4
Modifies the length of the std::string to the dcount+1 size (to accomodate for the trailing null-termiantion character)\&. 
.PP
If the resulting data size is less than dcount (like from EOF), resizes it to the amount read\&. 
.PP
If the file is closed, it doesn't modify the string
.RE
.PP
\fBParameters\fP
.RS 4
\fIstr\fP The refernce to the std::string object to read the bytes of string to
.br
\fIdcount\fP The amount of characters to read
.RE
.PP
\fBReturns\fP
.RS 4
AEFR_ERR_READ_SUCCESS on success; otherwise AEFR_ERR_FILE_NOT_OPEN if file isn't open, or other AEFR_ERR_READ_* flags (like AEFR_ERR_READ_EOF)
.RE
.PP

.PP
Definition at line \fB182\fP of file \fBAEFileReader\&.hpp\fP\&.
.PP
References \fB_AEFR_EXIT_ON_READ_CLOSED_FILE\fP, \fBm_szLastReadAmount\fP, and \fBreadString()\fP\&.
.SS "\fBcint\fP AEFileReader::readString (std::vector< char > & str, \fBconst\fP std::size_t dcount)\fR [inline]\fP"

.PP
Reads the given amount of bytes and interprets them as the string\&. 
.PP
\fBNote\fP
.RS 4
Modifies the length of the std::vector<char> to the dcount+1 size (to accomodate for the trailing null-termiantion character)\&. 
.PP
If the resulting data size is less than dcount (like from EOF), resizes it to the amount read (+1 for the null termination)\&. 
.PP
If the file is closed, it doesn't modify the std::vector<char>
.RE
.PP
\fBParameters\fP
.RS 4
\fIstr\fP The refernce to the std::vector<char> object to read the bytes of string to
.br
\fIdcount\fP The amount of characters to read
.RE
.PP
\fBReturns\fP
.RS 4
AEFR_ERR_READ_SUCCESS on success; otherwise AEFR_ERR_FILE_NOT_OPEN if file isn't open, or other AEFR_ERR_READ_* flags (like AEFR_ERR_READ_EOF)
.RE
.PP

.PP
Definition at line \fB201\fP of file \fBAEFileReader\&.hpp\fP\&.
.PP
References \fB_AEFR_EXIT_ON_READ_CLOSED_FILE\fP, \fBm_szLastReadAmount\fP, and \fBreadString()\fP\&.
.SS "\fBcint\fP AEFileReader::readString (char * str, \fBconst\fP std::size_t dcount)\fR [noexcept]\fP"

.PP
Reads the given amount of bytes and interprets them as the string\&. 
.PP
\fBNote\fP
.RS 4
The passed string must be at least dcount+1 characters long (+1 is for the trailing null termination) 
.PP
If the resulting data size is less than dcount (like from EOF), fills the rest of unfilled characters with NULL 
.PP
If the file is closed, it doesn't modify the data
.RE
.PP
\fBParameters\fP
.RS 4
\fIstr\fP The pointer to the c-string to read the bytes of string to
.br
\fIdcount\fP The amount of characters to read
.RE
.PP
\fBReturns\fP
.RS 4
AEFR_ERR_READ_SUCCESS on success; otherwise AEFR_ERR_FILE_NOT_OPEN if file isn't open, or other AEFR_ERR_READ_* flags (like AEFR_ERR_READ_EOF)
.RE
.PP

.PP
Definition at line \fB58\fP of file \fBAEFileReader\&.cpp\fP\&.
.PP
References \fB_AEFR_EXIT_ON_READ_CLOSED_FILE\fP, \fBAEFR_ERR_READ_ZERO_SIZE\fP, and \fBAENULL\fP\&.
.SS "\fBcint\fP AEFileReader::readStringNL (std::string & str, \fBconst\fP \fBint\fP dcount)\fR [inline]\fP"

.PP
Reads the characters to a std::string untill a newline (included), or untill the maximum character amount -- whichever comes first\&. 
.PP
\fBNote\fP
.RS 4
Modifies the length of the std::string to the dcount+1 size (to accomodate for the trailing null-termiantion character)\&. 
.PP
If the resulting data size is less than dcount (like from EOF or met newline), resizes it to the amount read\&. 
.PP
If the file is closed, it doesn't modify the string
.RE
.PP
\fBParameters\fP
.RS 4
\fIstr\fP The refernce to the std::string object to read the bytes of string to
.br
\fIdcount\fP The maximum amount of characters to read
.RE
.PP
\fBReturns\fP
.RS 4
AEFR_ERR_READ_SUCCESS on success; otherwise AEFR_ERR_FILE_NOT_OPEN if file isn't open, or other AEFR_ERR_READ_* flags (like AEFR_ERR_READ_EOF)
.RE
.PP

.PP
Definition at line \fB234\fP of file \fBAEFileReader\&.hpp\fP\&.
.PP
References \fB_AEFR_EXIT_ON_READ_CLOSED_FILE\fP, \fBm_szLastReadAmount\fP, and \fBreadStringNL()\fP\&.
.SS "\fBcint\fP AEFileReader::readStringNL (std::vector< char > & str, \fBconst\fP \fBint\fP dcount)\fR [inline]\fP"

.PP
Reads the characters to a std::vector<char> untill a newline (included), or untill the maximum character amount -- whichever comes first\&. 
.PP
\fBNote\fP
.RS 4
Modifies the length of the std::vector<char> to the dcount+1 size (to accomodate for the trailing null-termiantion character)\&. 
.PP
If the resulting data size is less than dcount (like from EOF or met newline), resizes it to the amount read (+1 for the null termination)\&. 
.PP
If the file is closed, it doesn't modify the std::vector<char>
.RE
.PP
\fBParameters\fP
.RS 4
\fIstr\fP The refernce to the std::vector<char> object to read the bytes of string to
.br
\fIdcount\fP The maximum amount of characters to read
.RE
.PP
\fBReturns\fP
.RS 4
AEFR_ERR_READ_SUCCESS on success; otherwise AEFR_ERR_FILE_NOT_OPEN if file isn't open, or other AEFR_ERR_READ_* flags (like AEFR_ERR_READ_EOF)
.RE
.PP

.PP
Definition at line \fB251\fP of file \fBAEFileReader\&.hpp\fP\&.
.PP
References \fB_AEFR_EXIT_ON_READ_CLOSED_FILE\fP, \fBm_szLastReadAmount\fP, and \fBreadStringNL()\fP\&.
.SS "\fBcint\fP AEFileReader::readStringNL (char * str, \fBconst\fP \fBint\fP dcount)\fR [noexcept]\fP"

.PP
Reads the characters to a c-string untill a newline (included), or untill the maximum character amount -- whichever comes first\&. 
.PP
\fBNote\fP
.RS 4
The passed string must be at least dcount+1 characters long (+1 is for the trailing null termination) 
.PP
If the resulting data size is less than dcount (like from EOF or met newline), fills the rest of unfilled characters with NULL 
.PP
If the file is closed, it doesn't modify the data
.RE
.PP
\fBParameters\fP
.RS 4
\fIstr\fP The pointer to the c-string to read the bytes of string to
.br
\fIdcount\fP The maximum amount of characters to read
.RE
.PP
\fBReturns\fP
.RS 4
AEFR_ERR_READ_SUCCESS on success; otherwise AEFR_ERR_FILE_NOT_OPEN if file isn't open, or other AEFR_ERR_READ_* flags (like AEFR_ERR_READ_EOF)
.RE
.PP

.PP
Definition at line \fB70\fP of file \fBAEFileReader\&.cpp\fP\&.
.PP
References \fB_AEFR_EXIT_ON_READ_CLOSED_FILE\fP, \fBAEFR_ERR_READ_EOF\fP, \fBAEFR_ERR_READ_ERROR\fP, \fBAEFR_ERR_READ_SUCCESS\fP, \fBAEFR_ERR_READ_ZERO_SIZE\fP, and \fBAENULL\fP\&.
.SS "\fBcint\fP AEFileReader::readStringNULL (std::string & str, \fBconst\fP \fBllint\fP dcount)\fR [inline]\fP"

.PP
Reads the characters to a std::string untill a NULL character(I guess, you can say 'included'?), or untill the maximum character amount -- whichever comes first\&. 
.PP
\fBNote\fP
.RS 4
Modifies the length of the std::string to the dcount+1 size (to accomodate for the trailing null-termiantion character)\&. 
.PP
If the resulting data size is less than dcount (like from EOF or met null), resizes it to the amount read\&. 
.PP
If the file is closed, it doesn't modify the string
.RE
.PP
\fBParameters\fP
.RS 4
\fIstr\fP The refernce to the std::string object to read the bytes of string to
.br
\fIdcount\fP The maximum amount of characters to read
.RE
.PP
\fBReturns\fP
.RS 4
AEFR_ERR_READ_SUCCESS on success; otherwise AEFR_ERR_FILE_NOT_OPEN if file isn't open, or other AEFR_ERR_READ_* flags (like AEFR_ERR_READ_EOF)
.RE
.PP

.PP
Definition at line \fB281\fP of file \fBAEFileReader\&.hpp\fP\&.
.PP
References \fB_AEFR_EXIT_ON_READ_CLOSED_FILE\fP, \fBm_szLastReadAmount\fP, and \fBreadStringNULL()\fP\&.
.SS "\fBcint\fP AEFileReader::readStringNULL (std::vector< char > & str, \fBconst\fP \fBllint\fP dcount)\fR [inline]\fP"

.PP

AEFileReader/src/AEFileReader.cpp

/*! @file AEFileReader/src/AEFileReader.cpp
 *  This file contains the definitions of the engine's file reader's functions
 *  @see AEFileReader
 *  @see AEFileReader.hpp
 *
 *  Should not cause everything to break :)
 */

#include "../include/AEFileReader.hpp"



AEFileReader::AEFileReader(const std::string_view fname) :
	m_sFilename(fname), m_ullTotalReads(0), m_szLastReadAmount(0), m_fpFilestr(nullptr) {

	this->openFile(fname);
}

AEFileReader::AEFileReader(AEFileReader&& fr) noexcept :
	m_sFilename(std::move(fr.m_sFilename)), m_ullTotalReads(fr.m_ullTotalReads.load()),
	m_szLastReadAmount(fr.m_szLastReadAmount), m_fpFilestr(fr.m_fpFilestr) {

	fr.m_sFilename.clear();
	fr.m_ullTotalReads = 0;
	fr.m_szLastReadAmount = 0;
	fr.m_fpFilestr = nullptr;
}



cint AEFileReader::openFile(const std::string_view fname) {
	
	if (this->isOpen()) {
		return AEFR_ERR_OPEN_FILE_ALREADY_OPENED;
	}


	if (fname.empty()) {
		return AEFR_ERR_OPEN_FILE_NAME_EMPTY;
	}

	this->m_szLastReadAmount = 0;
	this->m_fpFilestr = ace::utils::fopenCC(fname.data(), "rb");

	if (this->isClosed()) {
		// this can also be triggered if program has no permission to access the file
		// but generally it just would be the absence of file
		return AEFR_ERR_OPEN_FILE_DOESNT_EXIST;
	}

	this->m_sFilename = fname;
	return AEFR_ERR_NOERROR;
}

//reads the string of set size, Unfilled characters (if eof) are zeroed. MUST be of the dcount+1 size!
//if file is closed, contents is not modified
cint AEFileReader::readString(char* str, const std::size_t dcount) noexcept {
	_AEFR_EXIT_ON_READ_CLOSED_FILE;

	if (!dcount || !str) {
		return AEFR_ERR_READ_ZERO_SIZE;
	}

	std::memset(str, AENULL, dcount + 1);
	return this->readData_ptr(str, dcount, sizeof(char));
}

//reads the string untill the newline or, as a safety measure, untill dcount. MUST be of the dcount+1 size!
cint AEFileReader::readStringNL(char* str, const int dcount) noexcept {
	_AEFR_EXIT_ON_READ_CLOSED_FILE;

	if (!str || !dcount) {
		return AEFR_ERR_READ_ZERO_SIZE;
	}

	std::memset(str, AENULL, std::size_t(dcount) + 1);
	const bool temp = std::fgets(str, dcount, this->m_fpFilestr);
	this->m_szLastReadAmount = std::strlen(str);
	this->m_ullTotalReads.fetch_add(1, std::memory_order::relaxed);

	if (!temp) {
		return AEFR_ERR_READ_ERROR;
	}

	if (std::feof(this->m_fpFilestr)) {
		return AEFR_ERR_READ_EOF;
	}
	return AEFR_ERR_READ_SUCCESS;
}

//reads the string untill the null, or untill dcount
//moves the cursor to the found null char+1
cint AEFileReader::readStringNULL(char* str, const llint dcount) noexcept {
	_AEFR_EXIT_ON_READ_CLOSED_FILE;

	if (!str || dcount < 1) {
		return AEFR_ERR_READ_ZERO_SIZE;
	}

	//clear buffer and read
	//std::memset(str, AENULL, dcount + 1);
	str[dcount] = '\0'; //set the 0 to the last character (dcount+1) (just in case user didnt)
	const cint temp = this->readData_ptr(str, dcount, sizeof(char));
	this->m_szLastReadAmount = std::strlen(str);
	const llint differenceRead = dcount - llint(this->m_szLastReadAmount); // it's in bounds of the llint dcount anyway;
	//fill the rest of the buffer with 0's
	std::memset(str + this->m_szLastReadAmount, AENULL, differenceRead);

	//look for the null char and move the read cursor to the next char after null (if appeared)
	//const char* const nulchar = std::strchr(str, '\0');
	fseek(m_fpFilestr, -differenceRead, SEEK_CUR);
	

	return temp;
}

cint AEFileReader::readBoolString(bool& num) noexcept {
	_AEFR_EXIT_ON_READ_CLOSED_FILE;
	char str[sizeof("false")]{};

	cint temp = this->readData_ptr(str, 4, sizeof(char)); //read possible "true"
	for (int i = 0; i < 4; i++) {
		str[i] = std::tolower(str[i]);
	}

	if (!std::memcmp(str, "true", sizeof("true"))) {
		num = true;
		return temp;
	}

	temp = this->readData_ptr(&str[4], 1, sizeof(char)); //read last letter to make up "false"
	if (!std::memcmp(str, "false", sizeof("false"))) {
		num = false;
		return temp;
	}
	return AEFR_ERR_READ_ERROR; //whoops, nothing to read!
}

// reads data to ptr
cint AEFileReader::readData_ptr(void* cdata, const std::size_t dcount, const std::size_t dsize) noexcept {
	_AEFR_EXIT_ON_READ_CLOSED_FILE;

	if (!cdata || !dcount || !dsize) {
		return AEFR_ERR_READ_ZERO_SIZE;
	}

	this->m_szLastReadAmount = fread(cdata, dsize, dcount, this->m_fpFilestr);
	this->m_ullTotalReads.fetch_add(1, std::memory_order::relaxed);
	if (this->m_szLastReadAmount != dcount) {
		if (std::feof(this->m_fpFilestr)) {
			return AEFR_ERR_READ_EOF;
		}
		if (std::ferror(this->m_fpFilestr)) {
			return AEFR_ERR_READ_ERROR;
		}
	}
	return AEFR_ERR_READ_SUCCESS;
}
